closed/src/java.base/share/classes/jdk/crypto/jniprovider/NativeCrypto.java

```diff
@@ -178,6 +188,47 @@ public static final boolean isTraceEnabled() {
         return traceEnabled;
     }
 
+    public static final boolean isOpenSSLFIPSVersion() {
+        return InstanceHolder.instance.isOpenSSLFIPS;
+    }
+
+    /**
+     * Check whether a native implementation is available in the loaded OpenSSL library.
+     * Note that, an algorithm could be unavailable due to options used to build the
+     * OpenSSL version utilized, or using a FIPS version that doesn't allow it.
+     *
+     * @param algorithm the algorithm checked
+     * @return whether a native implementation of the given crypto algorithm is available
+     */
+    public static final boolean isAlgorithmAvailable(String algorithm) {
+        boolean isAlgorithmAvailable = false;
+        if (isAllowedAndLoaded()) {
+            if (isOpenSSLFIPSVersion()) {
+                if (disallowedAlgosFIPS.contains(algorithm)) {
+                    isAlgorithmAvailable = false;
+                }
+            } else {
+                switch (algorithm) {
+                    case "MD5":
+                        isAlgorithmAvailable = isMD5Available();
+                        break;
+                    default:
+                        isAlgorithmAvailable = true;
+                }
+            }
```

Comment:
This is still not right. In FIPS mode, `isAlgorithmAvailable` will always be `false`.
Suggest we can dispense with `disallowedAlgosFIPS` and just use a `switch` statement for FIPS and non-FIPS modes:
```java
            if (isOpenSSLFIPSVersion()) {
                switch (algorithm) {
                case "ChaCha20":
                case "MD5":
                    // not available
                    break;
                default:
                    isAlgorithmAvailable = true;
                    break;
                }
            } else {
                switch (algorithm) {
                case "MD5":
                    isAlgorithmAvailable = isMD5Available();
                    break;
                default:
                    isAlgorithmAvailable = true;
                    break;
                }
            }
```
Also notice `case` labels are indented the same as `switch` statements.

Reply:
Changed to suggestion.